src/qmapshack/gis/rte/router/CRouterBRouter.cpp

```diff
@@ -413,6 +413,33 @@ int CRouterBRouter::synchronousRequest(const QVector<QPointF> &points, const QLi
                         point.setX(elem.attribute("lon").toFloat()*DEG_TO_RAD);
                         point.setY(elem.attribute("lat").toFloat()*DEG_TO_RAD);
                     }
+
+                    //find costs of route (copied and adapted from CGisItemRte::setResultFromBrouter)
+                    if(costs != nullptr)
+                    {
+                        const QDomNodeList &nodes = xml.childNodes();
+                        for (int i = 0; i < nodes.count(); i++)
+                        {
+                            const QDomNode &node = nodes.at(i);
+                            if (node.isComment())
+                            {
+                                const QString &commentTxt = node.toComment().data();
+                                // ' track-length = 180864 filtered ascend = 428 plain-ascend = -172 cost=270249 '
+                                const QRegExp rxAscDes("(\\s*track-length\\s*=\\s*)(-?\\d+)(\\s*)(filtered ascend\\s*=\\s*-?\\d+)(\\s*)(plain-ascend\\s*=\\s*-?\\d+)(\\s*)(cost\\s*=\\s*)(-?\\d+)(\\s*)");
+                                int pos = rxAscDes.indexIn(commentTxt);
+                                if (pos > -1)
+                                {
+                                    bool ok;
+                                    *costs = rxAscDes.cap(9).toDouble(&ok);
+                                    if(!ok)
+                                    {
+                                        *costs = -1;
+                                    }
+                                }
+                                break;
+                            }
+                        }
+                    }
```

Comment:
Way too many nested code. Try to avoid that many hierarchies by escape logic or subfunctions.